src/lib/optimizer/strategy/predicate_reordering_rule.cpp

```diff
@@ -55,8 +55,10 @@ void PredicateReorderingRule::_apply_to_plan_without_subqueries(
   DebugAssert(cost_estimator, "PredicateReorderingRule requires cost estimator to be set");
   Assert(lqp_root->type == LQPNodeType::Root, "PredicateReorderingRule needs root to hold onto");
 
+  // We keep track of the visited predicate chains' nodes, so that this rule touches involved nodes only once.
+  std::unordered_set<std::shared_ptr<AbstractLQPNode>> visited_predicate_chain_nodes;
```

Comment:
Minor: `visited` means, in this specific case, that the corresponding nodes were already re-ordered. So we can be even more specific with the naming. How about `reordered_predicate_nodes`? However, this is not a merge stopper for me.